### PR TITLE
Improve upgrade tag discovery

### PR DESCRIPTION
**Pull request by knifewine**
give better info about tags/heads being used
allow finding 'tentative' tags
change trunk to 3.0
update version from strings to tuples for easier handling
query tags direct from apache git, or local if configured

### CODE REVIEW

**Comment by aboudreault** [comment]
I am not familiar with that upgrade test yet, but this LGTM!

**Comment by knifewine** [comment]
@ptnapoleon and/or @EnigmaCurry can you have a look at this? Changes don't alter the guts of the upgrade process used in tests, just make some alterations to how versions are specified and what tags are matched (so we can find 'tentative' tags).

File: upgrade_through_versions_test.py [hunk 1]
```diff
@@ -115,7 +143,7 @@ def __init__(self, *args, **kwargs):
     @property
     def test_versions(self):
         # Murmur was not present until 1.2+
```
Comment (ptnapoleon) [inline upgrade_through_versions_test.py]: Do we still need this? I didn't think we were testing upgrades from 1.1.x anymore.

**Comment by ptnapoleon** [comment]
+1.The code is verbose, but not too confusing. The list of improvements is quite nice.

**Comment by EnigmaCurry** [comment]
TIL git ls-remote.

LGTM.

**Pull request merged**